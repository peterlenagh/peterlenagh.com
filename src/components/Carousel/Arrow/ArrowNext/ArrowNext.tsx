import React, { memo, useContext } from "react";

import Arrow from "../Arrow";
import CarouselContext from "../../carousel-context";
type ArrowNextProps = React.ComponentPropsWithoutRef<"button">;

const ArrowNext = ({ ...rest }: ArrowNextProps) => {
  const { next, last, infinite } = useContext(CarouselContext);
  const noNext = last && !infinite;
  // if (noNext) return null;
  return (
    <Arrow
      data-testid="arrow-next"
      text="next"
      onClick={next}
      disabled={noNext}
      {...rest}
    />
  );
};

export default memo(ArrowNext);
