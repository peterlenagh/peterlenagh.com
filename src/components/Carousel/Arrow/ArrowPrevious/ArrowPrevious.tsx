import React, { memo, useContext } from "react";

import Arrow from "../Arrow";
import CarouselContext from "../../carousel-context";
type ArrowPreviousProps = React.ComponentPropsWithoutRef<'button'>;

const ArrowPrevious = ({ ...rest }: ArrowPreviousProps) => {
  const { prev, first, infinite } = useContext(CarouselContext);
  const noPrev = first && !infinite;
  // if (noPrev) return null;
  return (
    <Arrow
      data-testid="arrow-previous"
      text="previous"
      direction="left"
      onClick={prev}
      disabled={noPrev}
      {...rest}
    />
  );
};

export default memo(ArrowPrevious);
