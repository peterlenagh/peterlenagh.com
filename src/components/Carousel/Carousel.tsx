import React, {
  Children,
  useContext,
  useEffect,
  useLayoutEffect,
  useRef,
  useState,
} from "react";
import classnames from "classnames/bind";

import useSize from "@/utils/hooks/useSize";
import useScrollPosition from "@/utils/hooks/useScrollPosition";

import CarouselContext from "./carousel-context";

import styles from "./Carousel.module.scss";
import { flushSync } from "react-dom";

const cx = classnames.bind(styles);

type CarouselProps = React.ComponentPropsWithoutRef<"div"> & {
  options?: {
    infinite?: boolean,
  },
  onSlide?: () => void,
};

const Carousel = ({
  className = "",
  children,
  options = {},
  onSlide = () => {},
}: CarouselProps) => {
  const {
    selectedIndex,
    setItemsLength,
    itemsLength,
    setInfinite,
    looping,
    goTo,
    scroll,
    setScroll,
  } = useContext(CarouselContext);
  const blockClassName = cx(className, "carousel");

  const { infinite = false } = options;
  useEffect(() => {
    if (setInfinite) setInfinite(infinite);
  }, [infinite, setInfinite]);

  const [itemizedChildren] = useState(
    React.Children.map(children, (child) => {
      return <li className={cx("carousel__item")}>{child}</li>;
    })
  );

  useEffect(() => {
    const newItemsLength = Children.toArray(children).length;
    if (newItemsLength !== itemsLength && setItemsLength) setItemsLength(newItemsLength);
  }, [children, itemsLength, setItemsLength]);


  const carouselRef = useRef<HTMLUListElement>(null);
  const carouselContainerRef = useRef<HTMLDivElement>(null);
  const { width: slideWidth} = useSize(carouselRef);
  const { x: scrollLeft } = useScrollPosition(carouselRef);

  /**
   * On scroll change (left), if we're not programatically scrolling (scroll), select the right slide in state but dont literally
   * 'scroll' it (bc we're manually scrolling in this context)
   */
  useEffect(() => {
    if (typeof scrollLeft !== "number" || typeof slideWidth !== "number")
      return;
    const nearestSlideStartPx = Math.ceil(scrollLeft * slideWidth) / slideWidth;
    const index = Math.round(nearestSlideStartPx / slideWidth);
    goTo({ index, scroll: false });
  }, [goTo, scrollLeft, slideWidth]);

  /**
   * We scroll the carousel in response to selectedIndex changing.
   */
  useEffect(() => {
    if (!scroll) return;
    const scrollPos: { left: number, behavior: "auto" | "smooth" } = {
      left: Math.round(selectedIndex * slideWidth * 1e2) / 1e2,
      behavior: looping ? "auto" : "smooth",
    }
    if (carouselRef.current) carouselRef.current.scrollTo(scrollPos);
    setTimeout(() => { setScroll(false); }, 1);
    onSlide();
  }, [looping, selectedIndex, slideWidth, scroll, setScroll, carouselRef, onSlide]);

  return (
    <div ref={carouselContainerRef} className={blockClassName}>
      <ul ref={carouselRef} data-testid="carousel-list" className={cx("carousel__list")}>
        {itemizedChildren}
      </ul>
    </div>
  );
};

export default Carousel;
