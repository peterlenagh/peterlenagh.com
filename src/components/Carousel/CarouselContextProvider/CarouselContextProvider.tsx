import React, { memo, useEffect, useMemo, useReducer, Dispatch } from "react";

import carouselReducer, { actionCreators, CarouselState, initialState } from "../carousel-reducer";
import CarouselContext from "../carousel-context";
import { AnyAction, Action, ActionCreatorsMapObject } from "@reduxjs/toolkit";

type CarouselContextProviderProps = React.PropsWithChildren<{
  onChange?: (state: CarouselState) => void;
}>;

type PreserveArgsOnFnMap<Type extends { [k: string]: (...args: any[]) => any }> = {
  [Key in keyof Type]: (...args: Parameters<Type[Key]>) => void;
};

export type Expand<T> = T extends (...args: infer A) => infer R
  ? (...args: Expand<A>) => Expand<R>
  : T extends infer O
  ? { [K in keyof O]: O[K] }
  : never;

export type ExpandRecursively<T> = T extends (...args: infer A) => infer R
  ? (...args: ExpandRecursively<A>) => ExpandRecursively<R>
  : T extends object
  ? T extends infer O
    ? { [K in keyof O]: ExpandRecursively<O[K]> }
    : never
  : T;



function wrapActionCreatorsWithDispatch<
  Map extends ActionCreatorsMapObject
>(
  actionCreators: Map,
  dispatch: Dispatch<any>
): ExpandRecursively<PreserveArgsOnFnMap<Map>> {

  const wrappedActionCreators = {} as any;

  for (const key in actionCreators) {
    const actionCreator = actionCreators[key];
    if (typeof actionCreator === 'function') {
      wrappedActionCreators[key] = (...args: Parameters<typeof actionCreator>) => {
        dispatch(actionCreator(...args));
      };
    }
  }
  return wrappedActionCreators;
};

const CarouselContextProvider = ({ children, onChange = () => {} }: CarouselContextProviderProps) => {

  const [state, dispatch] = useReducer(carouselReducer, initialState)
  // const boundActionCreators = bindActionCreators(actionCreators, dispatch);
  const dispatchActions = useMemo(() => wrapActionCreatorsWithDispatch(actionCreators, dispatch),[dispatch]);

  // const actions = useMemo(() => ({
  //   prev: (...args: Parameters<typeof actionCreators.prev>) => dispatch(actionCreators.prev(...args)),
  //   next: (...args: Parameters<typeof actionCreators.next>) => dispatch(actionCreators.next(...args)),
  //   setItemsLength: (...args: Parameters<typeof actionCreators.setItemsLength>) => dispatch(actionCreators.setItemsLength(...args)),
  //   goTo: (...args: Parameters<typeof actionCreators.goTo>) => dispatch(actionCreators.goTo(...args)),
  //   setInfinite: (...args: Parameters<typeof actionCreators.setInfinite>) => dispatch(actionCreators.setInfinite(...args)),
  //   setScroll: (...args: Parameters<typeof actionCreators.setScroll>) => dispatch(actionCreators.setScroll(...args)),
  // }), [dispatch]);

  const value = React.useMemo(() => {
    return { ...state, ...dispatchActions };
  }, [state, dispatchActions]);

  useEffect(() => {
    onChange(value);
  }, [value, onChange]);

  return (
    <CarouselContext.Provider value={value}>
      {children}
    </CarouselContext.Provider>
  );
};

export default memo(CarouselContextProvider);
