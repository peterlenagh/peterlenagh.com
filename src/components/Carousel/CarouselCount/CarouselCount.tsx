import React, { memo, useContext } from "react";
import classnames from "classnames/bind";

import CarouselContext from "../carousel-context";

import styles from "./CarouselCount.module.scss";

const cx = classnames.bind(styles);

const CarouselCount = () => {
  const { selectedIndex, itemsLength } = useContext(CarouselContext);

  const current = (selectedIndex || 0) + 1;
  if (!itemsLength) return null;
  return (
    <div data-testid="carousel-count" className={cx("carousel-count")}>
      {current}/{itemsLength}
    </div>
  );
};

export default memo(CarouselCount);
