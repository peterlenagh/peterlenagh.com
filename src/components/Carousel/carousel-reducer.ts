/* eslint-disable no-param-reassign */
import produce from "immer";
import { isEqual, set as _set, merge } from "lodash";
import { ActionCreator, ActionCreatorsMapObject, createAction } from '@reduxjs/toolkit';



type goToPayload = {
    index: number
    scroll: boolean
};

export type CarouselApi = {
  prev: () => void,
  next: () => void,
  goTo: (index: number | goToPayload) => void,
  setItemsLength: (length: number) => void,
  setInfinite: (value: boolean) => void,
  setScroll: (value: boolean) => void
}

export const actionCreators = {
  prev: createAction('prev', () => ({payload: undefined})),
  next: createAction('next', () => ({payload: undefined})),
  goTo: createAction<number | goToPayload, 'goTo'>('goTo'),
  setItemsLength: createAction<number, 'setItemsLength'>('setItemsLength'),
  setInfinite: createAction<boolean, 'setInfinite'>('setInfinite'),
  setScroll: createAction<boolean, 'setScroll'>('setScroll'),
};

let test: ActionCreatorsMapObject<CarouselAction>

test = actionCreators;



export type CarouselAction = ReturnType<typeof actionCreators[keyof typeof actionCreators]>;

export const initialState = {
  selectedIndex: 0,
  itemsLength: 0,
  infinite: false,
  scroll: false,
  first: true,
  last: false,
  looping: false,
  goingTo: undefined,
};

export type CarouselState = typeof initialState;


const carouselReducer = (state = initialState, action: CarouselAction) => {
  const newState = produce(state, (draft) => {
    const { selectedIndex, itemsLength } = state;
    _set(draft, "looping", false);
    let newIndex = selectedIndex;
    let scroll = false;
    // eslint-disable-next-line default-case
    switch (action.type) {
      case 'setItemsLength': {
        const length = action.payload;
        _set(draft, "itemsLength", length);
        break;
      }
      case 'prev': {
        newIndex = selectedIndex - 1;
        if (newIndex < 0) {
          newIndex = itemsLength - 1;
          _set(draft, "looping", true);
        }
        scroll = true;
        break;
      }
      case 'next': {
        newIndex = selectedIndex + 1;
        if (newIndex > itemsLength - 1) {
          newIndex = 0;
          _set(draft, "looping", true);
        }
        scroll = true;
        break;
      }
      case 'goTo': {
        if (typeof action.payload === "object") {
          ({ index: newIndex, scroll } = action.payload);
        } else {
          newIndex = +action.payload;
          scroll = true;
        }
        break;
      }
      case 'setInfinite': {
        const value = action.payload;
        _set(draft, "infinite", value);
        break;
      }
      case 'setScroll': {
        const value = !!action.payload;
        _set(draft, "scroll", value);
        break;
      }
    }

    if (
      ['goTo','prev','next'].includes(
        action.type
      )
    ) {
      // if goingTo is set already, reset it if this action index is the same (i.e. we have "got to")
      if (draft.goingTo === newIndex) {
        _set(draft, "goingTo", undefined);
      }
      // if scroll = true, set goingTo index
      if (scroll) {
        _set(draft, "goingTo", newIndex);
      }
      // let that "going to" win over "new index" until we get to to where we're going
      const actualIndex =
        typeof draft.goingTo === "number" ? draft.goingTo : newIndex;
      _set(draft, "selectedIndex", actualIndex);
      _set(draft, "scroll", scroll);
    }

    const relativeState = {
      first: draft.selectedIndex === 0,
      last: draft.selectedIndex === draft.itemsLength - 1,
    };
    merge(draft, relativeState);
  });

  // thought immer did this for you too but nope :/
  if (isEqual(newState, state)) {
    return state;
  }
  return newState;
};

export default carouselReducer;
