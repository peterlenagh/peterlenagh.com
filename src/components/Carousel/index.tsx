"use client";

import React from "react";

import CarouselContextProvider from "./CarouselContextProvider";

import Carousel from "./Carousel";
import Arrow from "./Arrow/Arrow";
import ArrowNext from "./Arrow/ArrowNext/ArrowNext";
import ArrowPrevious from "./Arrow/ArrowPrevious/ArrowPrevious";
import CarouselContext from "./carousel-context";
import CarouselCount from "./CarouselCount/CarouselCount";
import CarouselPagination from "./CarouselPagination/CarouselPagination";

const withCarouselContext = <P extends object>(
  WrappedComponent: React.ComponentType<P>
) => {
  const WithCarouselContext = (props: React.PropsWithChildren) => (
    <CarouselContextProvider>
      <WrappedComponent {...(props as P)} />
    </CarouselContextProvider>
  );
  return WithCarouselContext;
};

export default Carousel;

export {
  CarouselContextProvider,
  Arrow,
  ArrowNext,
  ArrowPrevious,
  CarouselContext,
  withCarouselContext,
  CarouselCount,
  CarouselPagination,
};
