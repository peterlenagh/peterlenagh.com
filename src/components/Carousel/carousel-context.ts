import React from "react";
import { CarouselApi, CarouselState, initialState } from "./carousel-reducer";

const CarouselContext = React.createContext({ ...initialState } as CarouselState & CarouselApi);

export default CarouselContext;
