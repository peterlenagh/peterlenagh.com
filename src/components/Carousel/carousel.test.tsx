import { expect, test } from "@jest/globals";
import { act, render, screen } from "@testing-library/react";
import ArrowNext from "./Arrow/ArrowNext/ArrowNext";
import ArrowPrevious from "./Arrow/ArrowPrevious/ArrowPrevious";
import Carousel from "./Carousel";
import CarouselContextProvider from "./CarouselContextProvider";
import CarouselCount from "./CarouselCount/CarouselCount";

const slides = [
    {
        type: 'image',
        src: 'https://placekitten.com/g/700/700',
        ratio: 1
    },
    {
        type: 'image',
        src: 'https://placekitten.com/g/800/800',
        ratio: 1
    },
    {
        type: 'image',
        src: 'https://placekitten.com/g/900/900',
        ratio: 1
    },
    {
        type: 'image',
        src: 'https://placekitten.com/g/1000/1000',
        ratio: 1
    }
];

Element.prototype.scrollTo = jest.fn();
type CarouselWControlsProps = {
    infinite?: boolean
};

const CarouselWControls = ({ infinite = false }: CarouselWControlsProps) => (
    <CarouselContextProvider>
        <Carousel options={{ infinite }}>
            {slides.map((element, index) => {
                return (
                    // eslint-disable-next-line @next/next/no-img-element
                    <img
                        id={`slide-${index}`}
                        key={`slide-${index}`}
                        className="carousel__image"
                        alt={`image #${index + 1}`}
                    />
                );
            })}
        </Carousel>
        <ArrowPrevious />
        <ArrowNext />
        <CarouselCount />

    </CarouselContextProvider>
);

test("renders nothing if no children", () => {
    const carousel = render(
        <CarouselContextProvider>
            <Carousel />
        </CarouselContextProvider>
    );
    const list = carousel.getByTestId("carousel-list");
    expect(list.children.length).toBe(0);
});


test("renders carousel with images", () => {
    const carousel = render(<CarouselWControls />);
    const list = carousel.getByTestId("carousel-list");
    expect(list.children.length).toBe(slides.length);
});

test("renders snapshot", () => {
    const carousel = render(<CarouselWControls />);
    expect(carousel).toMatchSnapshot();
});

// test("next button scrolls carousel", async () => {
//     const carousel = render(<CarouselWControls />);
//     const next = carousel.getByTestId("arrow-next");
//     act(() => {
//         next.click();
//     });
//     const count = carousel.getByTestId("carousel-count");
//     expect(count.textContent).toBe("2/4");
// });
