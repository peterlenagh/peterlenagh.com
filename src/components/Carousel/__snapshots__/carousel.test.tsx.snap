// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="carousel"
      >
        <ul
          class="carousel__list"
          data-testid="carousel-list"
        >
          <li
            class="carousel__item"
          >
            <img
              alt="image #1"
              class="carousel__image"
              id="slide-0"
            />
          </li>
          <li
            class="carousel__item"
          >
            <img
              alt="image #2"
              class="carousel__image"
              id="slide-1"
            />
          </li>
          <li
            class="carousel__item"
          >
            <img
              alt="image #3"
              class="carousel__image"
              id="slide-2"
            />
          </li>
          <li
            class="carousel__item"
          >
            <img
              alt="image #4"
              class="carousel__image"
              id="slide-3"
            />
          </li>
        </ul>
      </div>
      <button
        class="button arrow"
        data-testid="arrow-previous"
      >
        previous
      </button>
      <button
        class="button arrow"
        data-testid="arrow-next"
      >
        next
      </button>
      <div
        class="carousel-count"
        data-testid="carousel-count"
      >
        1
        /
        4
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="carousel"
    >
      <ul
        class="carousel__list"
        data-testid="carousel-list"
      >
        <li
          class="carousel__item"
        >
          <img
            alt="image #1"
            class="carousel__image"
            id="slide-0"
          />
        </li>
        <li
          class="carousel__item"
        >
          <img
            alt="image #2"
            class="carousel__image"
            id="slide-1"
          />
        </li>
        <li
          class="carousel__item"
        >
          <img
            alt="image #3"
            class="carousel__image"
            id="slide-2"
          />
        </li>
        <li
          class="carousel__item"
        >
          <img
            alt="image #4"
            class="carousel__image"
            id="slide-3"
          />
        </li>
      </ul>
    </div>
    <button
      class="button arrow"
      data-testid="arrow-previous"
    >
      previous
    </button>
    <button
      class="button arrow"
      data-testid="arrow-next"
    >
      next
    </button>
    <div
      class="carousel-count"
      data-testid="carousel-count"
    >
      1
      /
      4
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
